src/macaron/slsa_analyzer/analyzer.py

```diff
@@ -513,58 +603,144 @@ def add_component(
-            ):
-                raise DuplicateCmpError(f"{repo_path} is already analyzed.", context=existing_record.context)
-
-            repository = self.add_repository(req_branch, git_obj)
-
-            if repository:
-                purl = PackageURL(
-                    type=repository.type,
-                    namespace=repository.owner,
-                    name=repository.name,
-                    version=repository.commit_sha,
+            if git_obj:
+                # TODO: use both the repo URL and the commit hash to check.
+                if (
+                    existing_records
+                    and (existing_record := existing_records.get(git_url.get_remote_origin_of_local_repo(git_obj)))
+                    is not None
+                ):
+                    raise DuplicateCmpError(
+                        f"{analysis_target.repo_path} is already analyzed.", context=existing_record.context
+                    )
+
+                repository = self.add_repository(analysis_target.branch, git_obj)
+            else:
+                # We cannot prepare the repository even though we have successfully resolved the repository path for the
+                # software component. If this happens, we don't raise error and treat the software component as if it
+                # does not have any ``Repository`` attached to it.
+                repository = None
+
+        if not analysis_target.parsed_purl:
+            # If the PURL is not available. This will only mean that the user don't provide PURL but only provide the
+            # repository path for the software component. Therefore, the ``Repository`` instance is used to create a
+            # unique PURL for it.
+            if not repository:
+                raise PURLNotFoundError(
+                    f"The repository {analysis_target.repo_path} is not available and no PURL is provided from the user."
                 )
 
-        # If PURL is not found, raise an exception.
-        if not purl:
-            logger.debug(
-                "Failed to locate PURL for repo: %s at branch %s and commit %s", repo_path, req_branch, req_digest
+            repo_snapshot_purl = PackageURL(
+                type=repository.type,
+                namespace=repository.owner,
+                name=repository.name,
+                version=repository.commit_sha,
             )
-            raise PURLNotFoundError("Failed to locate a PURL identifier for the component.")
+            return Component(purl=repo_snapshot_purl.to_string(), analysis=analysis, repository=repository)
+
+        # If the PURL is available, we always create the software component with it whether the repository is
+        # available or not.
+        return Component(purl=analysis_target.parsed_purl.to_string(), analysis=analysis, repository=repository)
+
+    @staticmethod
+    def to_analysis_target(config: Configuration, available_domains: list[str]) -> AnalysisTarget:
```

Comment:
I'm not sure why the return type of this function is a new abstract, `AnalysisTarget`. I want to understand if this is really needed when looks like this function is validating the inputs (`config`). If the relevant parts of the `config` argument are valid, the `config` doesn't need to change. Otherwise, it's better to update the `config` update itself and use it as the only source of truth, and raise relevant exceptions if the input is invalid.

Reply:
I don't update the values back to the `Configuration` instances is because I don't want to mutate the original `Configuration` instance provided by the users or provided by the dependency resolver. There are a couple of reasons why I don't think it's a good idea:
1. Given the nature of our current Configuration implementation, it's very difficult to write strictly typed code because the values are located inside a dictionary, this dictionary is assigned as an instance attribute within `Configuration` class. (see implementation [here](https://github.com/oracle/macaron/blob/db208d57aab379e2ec12bf953004bff061a4d320/src/macaron/config/target_config.py#L32)). Therefore, the type checker couldn't resolve the value of these values (see [here](https://github.com/oracle/macaron/blob/db208d57aab379e2ec12bf953004bff061a4d320/src/macaron/slsa_analyzer/analyzer.py#L692-L695) where the type of the extracted value are resolved to `Any` -> I need to explicitly annotate it to be `str`, which is not desirable).
2. From my understanding, the `Configuration` instances store the values acquired from outside of `Analyzer` (either from the user or from the dependency resolver). Therefore, I don't think it's a good idea to mutate it. Note that we also log the `Configuration` values to a file (see [here](https://github.com/oracle/macaron/blob/386c085d9ad30dd9eb9b35aef0c19b91718e02f7/src/macaron/output_reporter/reporter.py#L121)). 
3. The final resolved values like the repository path or the PURL object parsed from the PURL string are provided to the analysis eventually so I think we need to separate it from user inputs (which is defined in the `Configuration` instances). In additions, the types could be different too (the PURL string from the user and the PURL object parsed using `PackageURL`).

Reply:
As discussed offline, it would be better to improve the `Configuration` class and make it more restricted with types, and validations and use that instead of introducing a new abstraction . But let's think about the final solution a little more.